Plutarch/Verify.hs

```diff
@@ -0,0 +1,277 @@
+{-# LANGUAGE AllowAmbiguousTypes #-}
+{-# LANGUAGE UndecidableInstances #-}
+{-# OPTIONS_GHC -Wno-unticked-promoted-constructors #-}
+{-# OPTIONS_GHC -Wredundant-constraints #-}
+
+module Plutarch.Verify (
+  PTryFrom (PExcess, ptryFrom),
+  pcheckType,
+  pcheckByteStr,
+  pcheckInt,
+  pcheckList,
+  pcheckMap,
+) where
+
+import Plutarch.Builtin (
+  PAsData,
+  PBuiltinList,
+  PBuiltinMap,
+  PBuiltinPair,
+  PData,
+  PIsData (pfromData),
+  pasByteStr,
+  pasInt,
+  pasList,
+  pasMap,
+  pdata,
+  pforgetData,
+  pfstBuiltin,
+  ppairDataBuiltin,
+  psndBuiltin,
+ )
+
+import Plutarch.ByteString (PByteString)
+import Plutarch.Integer (PInteger)
+import Plutarch.Internal.Other (
+  POpaque,
+  PType,
+  Term,
+  pcon,
+  perror,
+  pforce,
+  phoistAcyclic,
+  plam,
+  plet,
+  (#),
+  type (:-->),
+ )
+import Plutarch.Unit (PUnit (PUnit))
+
+import Plutarch.DataRepr.Internal (
+  PDataRecord,
+  PLabeledType ((:=)),
+  pdcons,
+  pdnil,
+ )
+
+import Plutarch.Bool (pif, (#==))
+
+import Plutarch.List (
+  phead,
+  pmap,
+  ptail,
+ )
+
+import Plutarch.Unsafe (punsafeBuiltin, punsafeCoerce)
+import qualified PlutusCore as PLC
+
+import Plutarch.TermCont (TermCont, tcont, unTermCont)
+
+{- |
+    Note: PAsData POpaque ~ PData
+-}
+
+----------------------- The class PTryFrom ----------------------------------------------
+
+{- |
+    This checks the datastructure for validity.
+    If you don't care about parts of the datastructure
+    don't verify those parts, just let it return a PData
+    instead
+    Be aware this might get really expensive, so only
+    use it if you cannot establish trust otherwise
+    (e.g. via only checking a part of your Data with
+    PTryFrom)
+-}
+class PTryFrom (a :: PType) (b :: PType) where
```

Comment:
We should try to note the laws for the type class.
`fst <$> ptryFrom x ≡ pure $ punsafeCoerce x` if `x` is indeed also in `b`.